server/model/monitor.js

```diff
@@ -348,25 +348,30 @@ class Monitor extends BeanModel {
                     bean.msg = dnsMessage;
                     bean.status = UP;
                 } else if (this.type === "push") {      // Type: Push
-                    const time = R.isoDateTime(dayjs.utc().subtract(this.interval, "second"));
-
-                    let heartbeatCount = await R.count("heartbeat", " monitor_id = ? AND time > ? ", [
-                        this.id,
-                        time
-                    ]);
-
-                    log.debug("monitor", "heartbeatCount" + heartbeatCount + " " + time);
-
-                    if (heartbeatCount <= 0) {
-                        // Fix #922, since previous heartbeat could be inserted by api, it should get from database
-                        previousBeat = await Monitor.getPreviousHeartbeat(this.id);
-
-                        throw new Error("No heartbeat in the time window");
-                    } else {
-                        // No need to insert successful heartbeat for push type, so end here
-                        retries = 0;
-                        this.heartbeatInterval = setTimeout(beat, beatInterval * 1000);
-                        return;
+                    log.debug(`[${this.name}] Checking monitor at ${dayjs().format("YYYY-MM-DD HH:mm:ss.SSS")}`);
+                    const bufferTime = 1000; // 1s buffer to accommodate clock differences
+                    // Fix #922, since previous heartbeat could be inserted by api, it should get from database
+                    previousBeat = await Monitor.getPreviousHeartbeat(this.id);
+                    //If the previous beat was nonexistent, down or pending we use the regular
+                    // beatInterval/retryInterval in the setTimeout further below
+                    if (previousBeat) {
+                        const msSinceLastBeat = dayjs.utc().valueOf() - dayjs.utc(previousBeat.time).valueOf();
+                        log.debug(`[${this.name}] msSinceLastBeat = ${msSinceLastBeat}`);
+                        if (previousBeat.status !== UP || msSinceLastBeat > beatInterval * 1000 + bufferTime) {
+                            throw new Error("No heartbeat in the time window");
+                        } else {
+                            let timeout = beatInterval * 1000 - msSinceLastBeat;
+                            if (timeout < 0) {
+                                timeout = bufferTime;
+                            } else {
+                                timeout += bufferTime;
+                            }
```

Comment:
The math here is not correct - for example - I apologize if this ends up confusing

beatInterval = 1 second

So assuming the ticks are every second interval begins at 10:00:00.000  - that should mean that pushes arriving at 10:00:00.000, 10:00:00.800 and 10:00:02:200 should work because the first one needs to arrive by 10:00:00.000 and the second by 10:00:01.000 and the third at 10:00:02.000 but the third one is valid because we are allowing for a "buffer" now of 1 second.

The math here however doesn't work - the second one arrives at 10:00:00.800 meaning that it has a timeout of 1 * 1000 - 800 (difference between the first two ticks) - that equals 200 which means line 367 sets the timeout to 1200ms which means at 10:00:02.000 we will mark it as down before the third push arrives within the correct allowed window (1400 ms later). 

The same type of issue applies to something like pushes at 10:00:00.000, 10:00:01.200 and 10:00:02.600 because of the logic of line 365 

I had believed that line 363 should be

`let timeout = beatInterval * 2 * 1000 - msSinceLastBeat  + bufferTime`

That gives us (going back to the ifrst set of pushes)

1 * 2 * 1000 - 800 + 1000 = 2200ms for the 10:00:00.800 push (which would be correct)
1 * 2 * 1000 - 1400 + 1000 = 1600ms for the 10:00:02.200 push (which is incorrect)

So at this pint I know the math is incorrect as is - but I do not have something that appears to work properly for all cases.

Reply:
> because the first one needs to arrive by 10:00:00.000 and the second by 10:00:01.000 and the third at 10:00:02.000

This is incorrect, and why your extra `2 *` doesn't help. The whole point of this PR is for the timer to start counting down from scratch each and every time the push is received. In other words, suppose your interval is 20 seconds, and your buffer is 1 second. After a push is received, you have 21 seconds from the time the push was received to receive another push. Once that second push is received, no matter when its received, you have another 21 seconds from receiving push #2 to receive push #3.

So going back to your example (1 second interval, 1 second buffer):
* Push received at 10:00:00.000 - next push must arrive by 10:00:02.000
* Push received at 10:00:00.800 - next push must arrive by 10:00:02.800
* Push received at 10:00:02.000 - next push must arrive by 10:00:04.000

The top post in this PR - https://github.com/louislam/uptime-kuma/pull/1428#issue-1184022964 -  explains this in more detail, and I wrote about the race condition and why its a problem in this issue: https://github.com/louislam/uptime-kuma/issues/1422#issuecomment-1080016064

Reply:
That's not the current logic though of this patch.

10:00:00.000 next push by 10:00:02.000
10:00:00.800 next push by 10:00:02.000 (line 363 + line 367)
10:00:02.000 no error I think because it accepts the 10:00:00:800 push a second time (line 360) - and it moves out to 10:00:03.000 as the next required push (line 365)
10:00:02.900 (change for purpose of this example) - seems to pass because we're inside 10:00:03.000 (line 360)

If it's desired to just add the interval to the last push - then 363 should just be

`let timeout = dayjs.utc(previousBeat.time).valueOf() + beatInterval * 1000 + bufferTime` 

and the if block goes away because it's not necessary.

I'm not saying the original issue doesn't exist or there is nothing to do here at all - I'm just pointing out that the logic doesn't flow quite right.

Reply:
Ah the trick is that this section of code you're referring to isn't called when you hit the push URL. It's called by the monitor (part of the server code), which can theoretically be called at any time. This PR is set up to make it set up such that most of the time, it ends up being called around one interval + one buffer time after the previous beat.

Going back to your previous example, with some slight modifications in timing to avoid race condition confusion
1. 10:00:00.000 - push called
2. 10:00:00.001 - monitor is called - msSinceLastBeat = 1 - timeout = 1999 ms - next call at 10:00:02.000
3. 10:00:00.800 - push called
4. 10:00:02.000 - monitor is called - msSinceLastBeat = 1200 - timeout = 800ms - next call at 10:00:02.800
5. 10:00:02.001 - push called
6. 10:00:02.800 - monitor is called - msSinceLastBeat = 799 - timeout = 1201ms - next call at 10:00:04.001

Note that at each point when the monitor is called, its checking if msSinceLastBeat > 2000 (1s interval + 1s buffer). If that happens, the check is marked down. In this example, it never happens, and the check stays up.

If the push call is regular (i.e. cron job), the monitor will get eventually synchronized to the push call, and the buffer time allows for any latency in the push caller.

Also worth mentioning that I did test this, and it works totally fine! If you want to try it, you can clone my fork and set up a debug environment. I added debug statements so you can see how all of the timeouts and heartbeat checks are working.

Reply:
The logic for the 4th item is not what the code does - the if statement returns true instead of false because 1200ms is greater than the 1000ms beatInterval so the monitor doesn't fire until 10:00:03.000 which means it misses the fail if the #5 push is not until 10:00:02.900

Reply:
> The logic for the 4th item is not what the code does - the if statement returns true instead of false because 1200ms is greater than the 1000ms beatInterval so the monitor doesn't fire until 10:00:03.000 which means it misses the fail if the https://github.com/louislam/uptime-kuma/issues/5 push is not until 10:00:02.900

You mean line 360? The code compares `msSinceLastBeat` to `beatInterval * 1000 + bufferTime`.

Reply:
:360 is `msSinceLastBeat > beatInterval * 1000 + bufferTime`, so it would be `1200 > 2000` which is false.